File: typescript/src/signing.ts
import { createHmac, timingSafeEqual } from "node:crypto";

export const MOBIUS_SIGNATURE_HEADER = "X-Mobius-Signature";
export const MOBIUS_SIGNATURE_VERSION_HEADER = "X-Mobius-Signature-Version";
export const MOBIUS_TIMESTAMP_HEADER = "X-Mobius-Timestamp";
export const MOBIUS_DELIVERY_ID_HEADER = "X-Mobius-Delivery-Id";
export const MOBIUS_SECRET_REF_HEADER = "X-Mobius-Secret-Ref";
export const MOBIUS_SECRET_VERSION_HEADER = "X-Mobius-Secret-Version";

const SIGNATURE_PREFIX = "sha256=";
const VERSION = "v1";
const DEFAULT_MAX_AGE_SECONDS = 300;

export interface DeliveryMeta {
  signatureVersion: string;
  signature: string;
  timestamp: number;
  deliveryId: string;
  secretRef: string;
  secretVersion: number;
}

export interface VerifiedDelivery extends DeliveryMeta {
  body: Uint8Array;
}

export type ActionPrincipalType = "human" | "agent" | "service" | "system";
export type ActionInvocationOriginKind =
  | "agent_tool_call"
  | "loop_action_step"
  | "direct_action_invoke"
  | "server_internal";

export interface ActionInvocationScopeV1 {
  orgId: string;
  projectId: string;
}

export interface ActionInvocationActionV1 {
  id: string;
  name: string;
}

export interface ActionInvocationActorV1 {
  principalId: string;
  principalType: ActionPrincipalType;
  agentId?: string;
}

export interface ActionInvocationOriginV1 {
  kind: ActionInvocationOriginKind;
  runId?: string;
  channelExchangeId?: string;
  loopId?: string;
  stepKey?: string;
  agentTurnId?: string;
  sessionId?: string;
  toolCallId?: string;
}

export interface ActionInvocationContextV1 {
  schemaVersion: 1;
  scope: ActionInvocationScopeV1;
  action: ActionInvocationActionV1;
  actor: ActionInvocationActorV1;
  origin: ActionInvocationOriginV1;
}

export interface ActionInvocationV1 {
  mobius: ActionInvocationContextV1;
  parameters: Record<string, unknown>;
}

export interface VerifiedActionInvocationV1 extends VerifiedDelivery {
  invocation: ActionInvocationV1;
}

export type VerifySignedDeliveryOptions =
  | { key: Uint8Array; maxAge?: number; now?: () => number }
  | {
      resolveKey: (meta: DeliveryMeta) => Promise<Uint8Array> | Uint8Array;
      maxAge?: number;
      now?: () => number;
    };

export class InvalidSignatureError extends Error {
  constructor(message: string) {
    super(`mobius: invalid signed delivery: ${message}`);
    this.name = "InvalidSignatureError";
  }
}

export class StaleDeliveryError extends InvalidSignatureError {
  constructor(message: string) {
    super(message);
    this.name = "StaleDeliveryError";
  }
}

export class UnsupportedActionInvocationSchemaError extends Error {
  constructor(schemaVersion: unknown) {
    super(
      `mobius: unsupported action invocation schema: ${String(schemaVersion)}`,
    );
    this.name = "UnsupportedActionInvocationSchemaError";
  }
}

export class MalformedActionInvocationError extends Error {
  constructor(message: string) {
    super(`mobius: malformed action invocation: ${message}`);
    this.name = "MalformedActionInvocationError";
  }
}

export function readDeliveryMeta(request: Request): DeliveryMeta {
  return readDeliveryMetaFromHeaders(request.headers);
}

export function readDeliveryMetaFromHeaders(input: HeadersInit): DeliveryMeta {
  const headers = new Headers(input);
  const signatureVersion = requiredHeader(
    headers,
    MOBIUS_SIGNATURE_VERSION_HEADER,
  );
  if (signatureVersion !== VERSION) {
    throw new InvalidSignatureError("unsupported signature version");
  }
  const timestamp = Number(requiredHeader(headers, MOBIUS_TIMESTAMP_HEADER));
  if (!Number.isInteger(timestamp) || timestamp <= 0) {
    throw new InvalidSignatureError("invalid timestamp");
  }
  const secretVersion = Number(
    requiredHeader(headers, MOBIUS_SECRET_VERSION_HEADER),
  );
  if (!Number.isInteger(secretVersion) || secretVersion <= 0) {
    throw new InvalidSignatureError("invalid secret version");
  }
  return {
    signatureVersion,
    signature: requiredHeader(headers, MOBIUS_SIGNATURE_HEADER),
    timestamp,
    deliveryId: requiredHeader(headers, MOBIUS_DELIVERY_ID_HEADER),
    secretRef: requiredHeader(headers, MOBIUS_SECRET_REF_HEADER),
    secretVersion,
  };
}

export async function verifySignedDelivery(
  request: Request,
  opts: VerifySignedDeliveryOptions,
): Promise<VerifiedDelivery> {
  const body = new Uint8Array(await request.arrayBuffer());
  return verifySignedDeliveryBytes(body, request.headers, opts);
}

export async function verifySignedDeliveryBytes(
  body: Uint8Array,
  headers: HeadersInit,
  opts: VerifySignedDeliveryOptions,
): Promise<VerifiedDelivery> {
  const meta = readDeliveryMetaFromHeaders(headers);
  verifyFreshness(meta, opts);
  const key = "key" in opts ? opts.key : await opts.resolveKey(meta);
  if (!key?.byteLength) {
    throw new InvalidSignatureError("signing key is required");
  }
  verifySignature(key, body, meta);
  return { ...meta, body };
}

export async function verifyActionInvocationV1(
  body: Uint8Array,
  headers: HeadersInit,
  opts: VerifySignedDeliveryOptions,
): Promise<VerifiedActionInvocationV1> {
  const verified = await verifySignedDeliveryBytes(body, headers, opts);
  return { ...verified, invocation: parseActionInvocationV1(verified) };
}

export function signDelivery(
  key: Uint8Array,
  body: Uint8Array,
  opts: { deliveryId: string; timestamp: number },
): { signature: string } {
  const canonical = Buffer.concat([
    Buffer.from(`${VERSION}.${opts.deliveryId}.${opts.timestamp}.`),
    Buffer.from(body),
  ]);
  return {
    signature: `${SIGNATURE_PREFIX}${createHmac("sha256", key)
      .update(canonical)
      .digest("hex")}`,
  };
}

export function parseWebhookDelivery<T>(v: VerifiedDelivery): {
  type: string;
  data: T;
} {
  return JSON.parse(Buffer.from(v.body).toString("utf8")) as {
    type: string;
    data: T;
  };
}

export function parseActionInvocation<T>(v: VerifiedDelivery): T {
  return JSON.parse(Buffer.from(v.body).toString("utf8")) as T;
}

export function parseActionInvocationV1(
  v: VerifiedDelivery,
): ActionInvocationV1 {
  let payload: unknown;
  try {
    payload = JSON.parse(Buffer.from(v.body).toString("utf8"));
  } catch (error) {
    throw new MalformedActionInvocationError(
      `invalid JSON: ${error instanceof Error ? error.message : String(error)}`,
    );
  }
  const root = requiredObject(payload, "body");
  const mobius = requiredObject(root.mobius, "mobius");
  if (mobius.schema_version === undefined) {
    throw new MalformedActionInvocationError(
      "mobius.schema_version is required",
    );
  }
  if (mobius.schema_version !== 1) {
    throw new UnsupportedActionInvocationSchemaError(mobius.schema_version);
  }
  const scope = requiredObject(mobius.scope, "mobius.scope");
  const action = requiredObject(mobius.action, "mobius.action");
  const actor = requiredObject(mobius.actor, "mobius.actor");
  const origin = requiredObject(mobius.origin, "mobius.origin");
  const parameters = requiredObject(root.parameters, "parameters");

  const principalType = requiredString(
    actor.principal_type,
    "mobius.actor.principal_type",
  );
  if (!isActionPrincipalType(principalType)) {
    throw new MalformedActionInvocationError(
      "mobius.actor.principal_type is invalid",
    );
  }
  const agentId = optionalString(actor.agent_id, "mobius.actor.agent_id");
  if (principalType === "agent" && agentId === undefined) {
    throw new MalformedActionInvocationError(
      "mobius.actor.agent_id is required for agent actors",
    );
  }
  if (principalType !== "agent" && agentId !== undefined) {
    throw new MalformedActionInvocationError(
      "mobius.actor.agent_id is only valid for agent actors",
    );
  }
  const originKind = requiredString(origin.kind, "mobius.origin.kind");
  if (!isActionInvocationOriginKind(originKind)) {
    throw new MalformedActionInvocationError("mobius.origin.kind is invalid");
  }

  return {
    mobius: {
      schemaVersion: 1,
      scope: {
        orgId: requiredString(scope.org_id, "mobius.scope.org_id"),
        projectId: requiredString(scope.project_id, "mobius.scope.project_id"),
      },
      action: {
        id: requiredString(action.id, "mobius.action.id"),
        name: requiredString(action.name, "mobius.action.name"),
      },
      actor: {
        principalId: requiredString(
          actor.principal_id,
          "mobius.actor.principal_id",
        ),
        principalType,
        ...(agentId === undefined ? {} : { agentId }),
      },
      origin: {
        kind: originKind,
        ...optionalOriginFields(origin),
      },
    },
    parameters,
  };
}

export function parseInteractionCallback<T>(v: VerifiedDelivery): T {
  return JSON.parse(Buffer.from(v.body).toString("utf8")) as T;
}

function verifyFreshness(
  meta: DeliveryMeta,
  opts: { maxAge?: number; now?: () => number },
): void {
  const now = opts.now?.() ?? Math.floor(Date.now() / 1000);
  const maxAge = opts.maxAge ?? DEFAULT_MAX_AGE_SECONDS;
  if (Math.abs(now - meta.timestamp) > maxAge) {
    throw new StaleDeliveryError("timestamp outside max age");
  }
}

function verifySignature(
  key: Uint8Array,
  body: Uint8Array,
  meta: DeliveryMeta,
): void {
  if (!meta.signature.startsWith(SIGNATURE_PREFIX)) {
    throw new InvalidSignatureError("missing sha256 prefix");
  }
  const raw = meta.signature.slice(SIGNATURE_PREFIX.length);
  if (!/^[0-9a-fA-F]+$/.test(raw) || raw.length % 2 !== 0) {
    throw new InvalidSignatureError("signature is not hex");
  }
  const got = Buffer.from(raw, "hex");
  const expected = Buffer.from(
    signDelivery(key, body, {
      deliveryId: meta.deliveryId,
      timestamp: meta.timestamp,
    }).signature.slice(SIGNATURE_PREFIX.length),
    "hex",
  );
  if (got.length !== expected.length || !timingSafeEqual(got, expected)) {
    throw new InvalidSignatureError("mismatch");
  }
}

function requiredHeader(headers: Headers, name: string): string {
  const value = headers.get(name);
  if (!value) {
    throw new InvalidSignatureError(`missing ${name}`);
  }
  return value;
}

function requiredObject(value: unknown, path: string): Record<string, unknown> {
  if (typeof value !== "object" || value === null || Array.isArray(value)) {
    throw new MalformedActionInvocationError(`${path} must be an object`);
  }
  return value as Record<string, unknown>;
}

function requiredString(value: unknown, path: string): string {
  if (typeof value !== "string" || !value.trim()) {
    throw new MalformedActionInvocationError(`${path} is required`);
  }
  return value;
}

function optionalString(value: unknown, path: string): string | undefined {
  if (value === undefined) return undefined;
  if (typeof value !== "string" || !value.trim()) {
    throw new MalformedActionInvocationError(
      `${path} must be a non-empty string`,
    );
  }
  return value;
}

function isActionPrincipalType(value: string): value is ActionPrincipalType {
  return ["human", "agent", "service", "system"].includes(value);
}

function isActionInvocationOriginKind(
  value: string,
): value is ActionInvocationOriginKind {
  return [
    "agent_tool_call",
    "loop_action_step",
    "direct_action_invoke",
    "server_internal",
  ].includes(value);
}

function optionalOriginFields(
  origin: Record<string, unknown>,
): Omit<ActionInvocationOriginV1, "kind"> {
  const fields: Omit<ActionInvocationOriginV1, "kind"> = {};
  const mappings = [
    ["run_id", "runId"],
    ["channel_exchange_id", "channelExchangeId"],
    ["loop_id", "loopId"],
    ["step_key", "stepKey"],
    ["agent_turn_id", "agentTurnId"],
    ["session_id", "sessionId"],
    ["tool_call_id", "toolCallId"],
  ] as const;
  for (const [wireName, fieldName] of mappings) {
    const value = optionalString(origin[wireName], `mobius.origin.${wireName}`);
    if (value !== undefined) fields[fieldName] = value;
  }
  return fields;
}

File: python/deepnoodle/mobius/signing.py
from __future__ import annotations

import hashlib
import hmac
import json
import time
from collections.abc import Callable, Mapping
from dataclasses import dataclass
from typing import Any

MOBIUS_SIGNATURE_HEADER = "X-Mobius-Signature"
MOBIUS_SIGNATURE_VERSION_HEADER = "X-Mobius-Signature-Version"
MOBIUS_TIMESTAMP_HEADER = "X-Mobius-Timestamp"
MOBIUS_DELIVERY_ID_HEADER = "X-Mobius-Delivery-Id"
MOBIUS_SECRET_REF_HEADER = "X-Mobius-Secret-Ref"
MOBIUS_SECRET_VERSION_HEADER = "X-Mobius-Secret-Version"

_SIGNATURE_PREFIX = "sha256="
_VERSION = "v1"
_DEFAULT_MAX_AGE_SECONDS = 300


class InvalidSignatureError(ValueError):
    """Raised when a Mobius signed delivery is missing or invalid."""


class StaleDeliveryError(InvalidSignatureError):
    """Raised when a validly shaped signed delivery is outside the freshness window."""


class UnsupportedActionInvocationSchemaError(ValueError):
    """Raised when an action invocation uses an unknown envelope schema."""


class MalformedActionInvocationError(ValueError):
    """Raised when a v1 action invocation is structurally invalid."""


@dataclass(frozen=True)
class DeliveryMeta:
    signature_version: str
    signature: str
    timestamp: int
    delivery_id: str
    secret_ref: str
    secret_version: int


@dataclass(frozen=True)
class VerifiedDelivery(DeliveryMeta):
    body: bytes


@dataclass(frozen=True)
class ActionInvocationScopeV1:
    org_id: str
    project_id: str


@dataclass(frozen=True)
class ActionInvocationActionV1:
    id: str
    name: str


@dataclass(frozen=True)
class ActionInvocationActorV1:
    principal_id: str
    principal_type: str
    agent_id: str | None = None


@dataclass(frozen=True)
class ActionInvocationOriginV1:
    kind: str
    run_id: str | None = None
    channel_exchange_id: str | None = None
    loop_id: str | None = None
    step_key: str | None = None
    agent_turn_id: str | None = None
    session_id: str | None = None
    tool_call_id: str | None = None


@dataclass(frozen=True)
class ActionInvocationContextV1:
    schema_version: int
    scope: ActionInvocationScopeV1
    action: ActionInvocationActionV1
    actor: ActionInvocationActorV1
    origin: ActionInvocationOriginV1


@dataclass(frozen=True)
class ActionInvocationV1:
    mobius: ActionInvocationContextV1
    parameters: dict[str, Any]


@dataclass(frozen=True)
class VerifiedActionInvocationV1(VerifiedDelivery):
    invocation: ActionInvocationV1


def read_delivery_meta(headers: Mapping[str, str]) -> DeliveryMeta:
    signature_version = _required_header(headers, MOBIUS_SIGNATURE_VERSION_HEADER)
    if signature_version != _VERSION:
        raise InvalidSignatureError("unsupported signature version")
    timestamp = _parse_positive_int(
        _required_header(headers, MOBIUS_TIMESTAMP_HEADER),
        "invalid timestamp",
    )
    secret_version = _parse_positive_int(
        _required_header(headers, MOBIUS_SECRET_VERSION_HEADER),
        "invalid secret version",
    )
    return DeliveryMeta(
        signature_version=signature_version,
        signature=_required_header(headers, MOBIUS_SIGNATURE_HEADER),
        timestamp=timestamp,
        delivery_id=_required_header(headers, MOBIUS_DELIVERY_ID_HEADER),
        secret_ref=_required_header(headers, MOBIUS_SECRET_REF_HEADER),
        secret_version=secret_version,
    )


def sign_delivery(
    key: bytes,
    body: bytes | str,
    *,
    delivery_id: str,
    timestamp: int,
) -> str:
    payload = _body_bytes(body)
    canonical = b".".join(
        [
            _VERSION.encode(),
            delivery_id.encode(),
            str(timestamp).encode(),
            payload,
        ]
    )
    signature = hmac.new(key, canonical, hashlib.sha256).hexdigest()
    return f"{_SIGNATURE_PREFIX}{signature}"


def verify_signed_delivery(
    body: bytes | str,
    headers: Mapping[str, str],
    *,
    key: bytes | None = None,
    resolve_key: Callable[[DeliveryMeta], bytes] | None = None,
    max_age: int = _DEFAULT_MAX_AGE_SECONDS,
    now: Callable[[], int] | None = None,
) -> VerifiedDelivery:
    payload = _body_bytes(body)
    meta = read_delivery_meta(headers)
    _verify_freshness(meta.timestamp, max_age=max_age, now=now)
    signing_key = key
    if signing_key is None and resolve_key is not None:
        signing_key = resolve_key(meta)
    if not signing_key:
        raise InvalidSignatureError("signing key is required")
    _verify_signature(signing_key, payload, meta)
    return VerifiedDelivery(
        signature_version=meta.signature_version,
        signature=meta.signature,
        timestamp=meta.timestamp,
        delivery_id=meta.delivery_id,
        secret_ref=meta.secret_ref,
        secret_version=meta.secret_version,
        body=payload,
    )


def verify_action_invocation_v1(
    body: bytes | str,
    headers: Mapping[str, str],
    *,
    key: bytes | None = None,
    resolve_key: Callable[[DeliveryMeta], bytes] | None = None,
    max_age: int = _DEFAULT_MAX_AGE_SECONDS,
    now: Callable[[], int] | None = None,
) -> VerifiedActionInvocationV1:
    """Verify exact raw bytes, then parse and validate the signed v1 envelope."""
    verified = verify_signed_delivery(
        body,
        headers,
        key=key,
        resolve_key=resolve_key,
        max_age=max_age,
        now=now,
    )
    invocation = parse_action_invocation_v1(verified)
    return VerifiedActionInvocationV1(
        signature_version=verified.signature_version,
        signature=verified.signature,
        timestamp=verified.timestamp,
        delivery_id=verified.delivery_id,
        secret_ref=verified.secret_ref,
        secret_version=verified.secret_version,
        body=verified.body,
        invocation=invocation,
    )


def parse_webhook_delivery(v: VerifiedDelivery) -> dict[str, Any]:
    payload = _parse_json_object(v, "webhook delivery")
    event_type = payload.get("type")
    if not isinstance(event_type, str) or not event_type:
        raise ValueError("mobius: webhook delivery missing 'type'")
    return payload


def parse_action_invocation(v: VerifiedDelivery) -> dict[str, Any]:
    return _parse_json_object(v, "action invocation")


def parse_action_invocation_v1(v: VerifiedDelivery) -> ActionInvocationV1:
    try:
        payload = _parse_json_object(v, "action invocation")
    except (UnicodeDecodeError, json.JSONDecodeError, TypeError, ValueError) as exc:
        if isinstance(exc, MalformedActionInvocationError):
            raise
        raise MalformedActionInvocationError("mobius: malformed action invocation: invalid JSON object") from exc

    mobius = _required_object(payload, "mobius")
    schema_version = mobius.get("schema_version")
    if schema_version is None:
        raise MalformedActionInvocationError("mobius: malformed action invocation: mobius.schema_version is required")
    if not isinstance(schema_version, int) or isinstance(schema_version, bool):
        raise MalformedActionInvocationError(
            "mobius: malformed action invocation: mobius.schema_version must be an integer"
        )
    if schema_version != 1:
        raise UnsupportedActionInvocationSchemaError(
            f"mobius: unsupported action invocation schema: {schema_version}"
        )

    scope = _required_object(mobius, "scope")
    action = _required_object(mobius, "action")
    actor = _required_object(mobius, "actor")
    origin = _required_object(mobius, "origin")
    parameters = _required_object(payload, "parameters")

    principal_type = _required_string(actor, "principal_type", "mobius.actor")
    if principal_type not in {"human", "agent", "service", "system"}:
        raise MalformedActionInvocationError(
            "mobius: malformed action invocation: mobius.actor.principal_type is invalid"
        )
    agent_id = actor.get("agent_id")
    if agent_id is not None and (not isinstance(agent_id, str) or not agent_id.strip()):
        raise MalformedActionInvocationError(
            "mobius: malformed action invocation: mobius.actor.agent_id must be a non-empty string"
        )
    if principal_type == "agent" and not agent_id:
        raise MalformedActionInvocationError(
            "mobius: malformed action invocation: mobius.actor.agent_id is required for agent actors"
        )
    if principal_type != "agent" and agent_id is not None:
        raise MalformedActionInvocationError(
            "mobius: malformed action invocation: mobius.actor.agent_id is only valid for agent actors"
        )

    origin_kind = _required_string(origin, "kind", "mobius.origin")
    if origin_kind not in {
        "agent_tool_call",
        "loop_action_step",
        "direct_action_invoke",
        "server_internal",
    }:
        raise MalformedActionInvocationError(
            "mobius: malformed action invocation: mobius.origin.kind is invalid"
        )

    return ActionInvocationV1(
        mobius=ActionInvocationContextV1(
            schema_version=1,
            scope=ActionInvocationScopeV1(
                org_id=_required_string(scope, "org_id", "mobius.scope"),
                project_id=_required_string(scope, "project_id", "mobius.scope"),
            ),
            action=ActionInvocationActionV1(
                id=_required_string(action, "id", "mobius.action"),
                name=_required_string(action, "name", "mobius.action"),
            ),
            actor=ActionInvocationActorV1(
                principal_id=_required_string(actor, "principal_id", "mobius.actor"),
                principal_type=principal_type,
                agent_id=agent_id,
            ),
            origin=ActionInvocationOriginV1(
                kind=origin_kind,
                run_id=_optional_string(origin, "run_id", "mobius.origin"),
                channel_exchange_id=_optional_string(origin, "channel_exchange_id", "mobius.origin"),
                loop_id=_optional_string(origin, "loop_id", "mobius.origin"),
                step_key=_optional_string(origin, "step_key", "mobius.origin"),
                agent_turn_id=_optional_string(origin, "agent_turn_id", "mobius.origin"),
                session_id=_optional_string(origin, "session_id", "mobius.origin"),
                tool_call_id=_optional_string(origin, "tool_call_id", "mobius.origin"),
            ),
        ),
        parameters=dict(parameters),
    )


def parse_interaction_callback(v: VerifiedDelivery) -> dict[str, Any]:
    return _parse_json_object(v, "interaction callback")


def _parse_json_object(v: VerifiedDelivery, kind: str) -> dict[str, Any]:
    payload = json.loads(v.body.decode())
    if not isinstance(payload, dict):
        raise ValueError(f"mobius: {kind} body must be a JSON object")
    return payload


def _verify_signature(key: bytes, body: bytes, meta: DeliveryMeta) -> None:
    if not meta.signature.startswith(_SIGNATURE_PREFIX):
        raise InvalidSignatureError("missing sha256 prefix")
    raw = meta.signature[len(_SIGNATURE_PREFIX) :]
    try:
        bytes.fromhex(raw)
    except ValueError as exc:
        raise InvalidSignatureError("signature is not hex") from exc
    expected = sign_delivery(
        key,
        body,
        delivery_id=meta.delivery_id,
        timestamp=meta.timestamp,
    )
    if not hmac.compare_digest(meta.signature, expected):
        raise InvalidSignatureError("mismatch")


def _verify_freshness(
    timestamp: int,
    *,
    max_age: int,
    now: Callable[[], int] | None,
) -> None:
    current = now() if now is not None else int(time.time())
    if abs(current - timestamp) > max_age:
        raise StaleDeliveryError("timestamp outside max age")


def _required_object(value: Mapping[str, Any], key: str) -> dict[str, Any]:
    candidate = value.get(key)
    if not isinstance(candidate, dict):
        raise MalformedActionInvocationError(
            f"mobius: malformed action invocation: {key} must be an object"
        )
    return candidate


def _required_string(value: Mapping[str, Any], key: str, path: str) -> str:
    candidate = value.get(key)
    if not isinstance(candidate, str) or not candidate.strip():
        raise MalformedActionInvocationError(
            f"mobius: malformed action invocation: {path}.{key} is required"
        )
    return candidate


def _optional_string(value: Mapping[str, Any], key: str, path: str) -> str | None:
    candidate = value.get(key)
    if candidate is None:
        return None
    if not isinstance(candidate, str) or not candidate.strip():
        raise MalformedActionInvocationError(
            f"mobius: malformed action invocation: {path}.{key} must be a non-empty string"
        )
    return candidate


def _required_header(headers: Mapping[str, str], key: str) -> str:
    for candidate, value in headers.items():
        if candidate.lower() == key.lower() and value:
            return value
    raise InvalidSignatureError(f"missing {key}")


def _parse_positive_int(raw: str, message: str) -> int:
    try:
        value = int(raw)
    except ValueError as exc:
        raise InvalidSignatureError(message) from exc
    if value <= 0:
        raise InvalidSignatureError(message)
    return value


def _body_bytes(body: bytes | str) -> bytes:
    return body.encode() if isinstance(body, str) else body
